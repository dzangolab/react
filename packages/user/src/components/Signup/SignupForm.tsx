import { Provider, emailSchema } from "@dzangolab/react-form";
import { useTranslation } from "@dzangolab/react-i18n";
import React from "react";
import * as zod from "zod";

import { PasswordConfirmationSchema } from "../schemas";
import SignupFormFields from "./SignupFormFields";
import { useConfig } from "../../hooks";

import type { LoginCredentials } from "../../types";

interface Properties {
  email?: string;
  loading?: boolean;
  termsAndConditions?: React.ReactNode;
  handleSubmit: (credentials: LoginCredentials) => void;
}

export const SignupForm = ({
  email,
  loading,
  termsAndConditions,
  handleSubmit,
}: Properties) => {
  const { t, i18n } = useTranslation("user");
  const config = useConfig();

  const SignUpFormSchema = zod
    .object({
      email: emailSchema({
        invalid: t("validation.messages.validEmail"),
        required: t("validation.messages.email"),
      }),
      ...PasswordConfirmationSchema({
        passwordRequiredMessage: t("signup.messages.validation.password"),
        passwordValidationMessage: t(
          "signup.messages.validation.validationMessage",
        ),
        confirmPasswordRequiredMessage: t(
          "signup.messages.validation.confirmPassword",
        ),
      }),
      ...(config.features?.termsAndConditions?.display &&
      config.features.termsAndConditions.showCheckbox
        ? {
            termsAndConditions: zod
              .boolean()
              .refine((value) => value === true, {
                message: t("signup.messages.validation.termsAndConditions"),
              }),
          }
        : {}),
    })
    .refine(
      (data) => {
        return data.password === data.confirmPassword;
      },
      {
        message: t("signup.messages.validation.mustMatch"),
        path: ["confirmPassword"],
      },
    );

  return (
    <Provider
      onSubmit={handleSubmit}
      defaultValues={{
        password: "",
        email: email || "",
        confirmPassword: "",
        termsAndConditions: false,
      }}
      validationSchema={SignUpFormSchema}
      validationTriggerKey={i18n.language}
    >
      <SignupFormFields
        disableEmailField={!!email}
        loading={loading}
        termsAndConditions={termsAndConditions}
      />
    </Provider>
  );
};
