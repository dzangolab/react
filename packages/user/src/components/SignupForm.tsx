import { useTranslation } from "@dzangolab/react-i18n";
import { LoadingButton } from "@dzangolab/react-ui";
import { Field, Formik } from "formik";
import React from "react";
import * as Yup from "yup";

import ErrorMessage from "./ErrorMessage";

import type { LoginCredentials } from "@/types/types";
import PasswordConfirmationWidget from "./PasswordConfirmationWidget";

interface Properties {
  handleSubmit: (credentials: LoginCredentials) => void;
  loading?: boolean;
}

const SignupForm = ({ handleSubmit, loading }: Properties) => {
  const { t } = useTranslation("user");

  const SingUpFormSchema = Yup.object({
    email: Yup.string()
      .email("validation.messages.validEmail")
      .required("validation.messages.email"),
    password: Yup.string()
      .matches(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d\w\W]{8,}$/,
        "signup.messages.validation.validationMessage"
      )
      .required("signup.messages.validation.password"),
    confirmPassword: Yup.string()
      .oneOf(
        [Yup.ref("password"), null],
        "signup.messages.validation.mustMatch"
      )
      .required("signup.messages.validation.confirmPassword"),
  });

  const initialValue = {
    email: "",
    password: "",
    confirmPassword: "",
  };

  return (
    <Formik
      initialValues={initialValue}
      validationSchema={SingUpFormSchema}
      onSubmit={(values, action) => {
        const data = { email: values.email, password: values.password };
        handleSubmit(data);
        action.resetForm();
      }}
    >
      {({ errors, handleSubmit, touched }) => (
        <form className="form" onSubmit={handleSubmit}>
          <div className="field email">
            <label htmlFor="email">{t("signup.form.email.label")}</label>
            <Field
              id="email"
              type="email"
              name="email"
              placeholder={t("signup.form.email.placeholder")}
            />
            <ErrorMessage
              touched={touched.email}
              error={errors.email ? t(errors.email) : undefined}
            />
          </div>

          {/* <div className="field password">
            <label htmlFor="password">{t("signup.form.password.label")}</label>
            <Field id="password" type="password" name="password" />
            <ErrorMessage
              touched={touched.password}
              error={errors.password ? t(errors.password) : undefined}
            />
          </div> */}

          {/* <div className="field password">
            <label htmlFor="confirmPassword">
              {t("signup.form.confirmPassword.label")}
            </label>
            <Field
              id="confirmPassword"
              type="password"
              name="confirmPassword"
            />
            <ErrorMessage
              touched={touched.confirmPassword}
              error={
                errors.confirmPassword ? t(errors.confirmPassword) : undefined
              }
            />
          </div> */}

          <PasswordConfirmationWidget errors={errors} touched={touched} />

          <div className="actions">
            <LoadingButton
              label={t("signup.form.actions.submit")}
              loading={loading}
            />
          </div>
        </form>
      )}
    </Formik>
  );
};

export default SignupForm;
