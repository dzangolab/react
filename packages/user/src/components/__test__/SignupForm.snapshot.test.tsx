import React from "react";
import renderer from "react-test-renderer";
import { expect, test } from "vitest";

import SignupForm from "../SignupForm";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const component = renderer.create(<SignupForm handleSubmit={() => {}} />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
