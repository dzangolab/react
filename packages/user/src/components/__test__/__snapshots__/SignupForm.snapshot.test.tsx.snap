// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<form
  className=""
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="field email"
  >
    <label
      htmlFor="email"
    >
      signup.form.email.label
    </label>
    <input
      disabled={false}
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="signup.form.email.placeholder"
      readOnly={false}
      type="email"
      value=""
    />
  </div>
  <div
    className="field password"
  >
    <label
      htmlFor="input-field-password"
    >
      signup.form.password.label
    </label>
    <div
      className="input-field-password"
    >
      <input
        id="input-field-password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
        type="password"
      />
      <span
        className="eye-icon"
        onClick={[Function]}
      >
        <i
          className="pi pi-eye-slash"
        />
      </span>
    </div>
  </div>
  <div
    className="field confirmPassword"
  >
    <label
      htmlFor="input-field-confirmPassword"
    >
      signup.form.confirmPassword.label
    </label>
    <div
      className="input-field-password"
    >
      <input
        id="input-field-confirmPassword"
        name="confirmPassword"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
        type="password"
      />
      <span
        className="eye-icon"
        onClick={[Function]}
      >
        <i
          className="pi pi-eye-slash"
        />
      </span>
    </div>
  </div>
  <div
    className="dz-form-actions"
    data-alignment="fill"
    data-direction="horizontal"
  >
    <button
      className="dz-button primary filled medium"
      id="submit"
      role="button"
      type="submit"
    >
      <span
        className="label"
      >
        signup.form.actions.submit
      </span>
    </button>
  </div>
</form>
`;
